### PR TITLE
Add all packages

### PR DESCRIPTION
**Pull request by knopp**

### CODE REVIEW

File: example/dart_package/README.md [hunk 1]
```diff
@@ -0,0 +1,39 @@
+<!--
```
Comment (mosuem) [inline example/dart_package/README.md:1]: An actual readme would be nice ;)

File: example/dart_package/hook/build.dart [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+    await build(args, (BuildConfig buildConfig, BuildOutput output) async {
+      final builder = RustBuilder(
+        package: 'dart_package',
+        crateManifestPath: 'rust/Cargo.toml',
+        buildConfig: buildConfig,
+      );
+      await builder.run(output: output);
+    });
+  } catch (e) {
+    print(e);
```
Comment (mosuem) [inline example/dart_package/hook/build.dart:17]: Why is this try-catch needed?

File: example/dart_package/lib/dart_package.dart [hunk 3]
```diff
@@ -0,0 +1,8 @@
+/// Support for doing something awesome.
```
Comment (mosuem) [inline example/dart_package/lib/dart_package.dart:1]: Clean up remains from `dart create` (also there is a `-e` flag to not add comments for new project creation)

File: native_doctor/README.md [hunk 4]
```diff
@@ -0,0 +1,93 @@
+# Native Doctor
```
Comment (mosuem) [inline native_doctor/README.md:1]: :rocket:

File: native_doctor/lib/src/native_doctor.dart [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+    if (manifests.isEmpty) {
+      writer.printMessage(
+        'No dependency containing ${NativeManifest.fileName} found. Nothing to check.',
+      );
+      writer.emptyLine();
+      return;
+    }
+
+    final checkers = [
+      NdkToolchainChecker(doctor: this),
```
Comment (mosuem) [inline native_doctor/lib/src/native_doctor.dart:138]: I think it would be even nicer to have this as an extensible mechanism per package - so that other packages could introduce their own checkers, without having to file a PR here. This could mean something like traversing the package dependencies and collecting top-level `native_doctor_check.dart` files defining a top-level method `ToolchainChecker checker(Doctor doctor)` or similar mechanisms. Otherwise, this is more of a `native_doctor_rust`.

File: native_doctor/lib/src/native_doctor.dart [hunk 6]
```diff
@@ -0,0 +1,320 @@
...
+      writer: writer,
+    );
+
+    if (options.path.existsSync()) {
+      final doctor = await NativeDoctor.withOptions(options);
+      doctor.run();
+    } else {
+      throw ToolError('Project path ${options.path.path} does not exist.');
+    }
+  } on FormatException catch (e) {
```
Comment (mosuem) [inline native_doctor/lib/src/native_doctor.dart:305]: Nit: Can this `try-catch` block be made smaller, for instance only wrapping `argParser.parse`? Large `try-catch` blocks make me nervous :)

File: native_doctor/lib/src/writer.dart [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+
+  void _timerCallback(Timer timer) {
+    _writeToStdOut('\b${_animation[_animationFrame]}');
+    _animationFrame = (_animationFrame + 1) % _animation.length;
+  }
+
+  void _beginSpinner() {
+    _spinerTimmer?.cancel();
+    if (hasTerminal) {
+      _animation = supportsEmoji ? '⠙⠚⠖⠦⢤⣠⣄⡤⠴⠲⠓⠋' : r'-\|/';
```
Comment (mosuem) [inline native_doctor/lib/src/writer.dart:155]: Nice details :)

File: native_doctor/pubspec.lock [hunk 8]
```diff
@@ -0,0 +1,459 @@
+# Generated by pub
```
Comment (mosuem) [inline native_doctor/pubspec.lock:1]: Do you want to check in the `pubspec.lock` for this?

File: example/dart_package/hook/build.dart [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+
+import 'package:native_toolchain_rust/native_toolchain_rust.dart';
+import 'package:native_assets_cli/native_assets_cli.dart';
+
+void main(List<String> args) async {
+  try {
+    await build(args, (BuildConfig buildConfig, BuildOutput output) async {
+      final builder = RustBuilder(
+        package: 'dart_package',
+        crateManifestPath: 'rust/Cargo.toml',
```
Comment (mosuem) [inline example/dart_package/hook/build.dart:11]: API-wise, why point to the `Cargo.toml` file itself and not just the folder?

File: native_toolchain_rust/lib/src/builder.dart [hunk 10]
```diff
@@ -0,0 +1,275 @@
...
+
+  final String message;
+
+  @override
+  String toString() {
+    return message;
+  }
+}
+
+class RunNativeDoctorException extends RustToolchainException {
```
Comment (mosuem) [inline native_toolchain_rust/lib/src/builder.dart:24]: Just a thought: does there need to be some versioning wrt `native doctor` and `native_toolchain_rust`? `dart pub global activate native_doctor` will always get the newest version, while the user may be using an old version of `native_toolchain_rust`. Probably not an issue though, as the information is passed via the native manifest `yaml`?

File: rustup/lib/src/rustup.dart [hunk 11]
```diff
@@ -0,0 +1,312 @@
...
+    ),
+    RustTarget(
+      os: OS.linux,
+      architecture: Architecture.x64,
+      triple: 'x86_64-unknown-linux-gnu',
+    ),
+  ];
+}
+
+class RustupToolchain {
```
Comment (mosuem) [inline rustup/lib/src/rustup.dart:266]: I personally like to have ~1 class per file, especially once the classes are larger than a few lines. Makes it easier to see what changed in a commit by seeing the file names only.

File: native_toolchain_rust_common/pubspec.yaml [hunk 12]
```diff
@@ -0,0 +1,23 @@
+name: native_toolchain_rust_common
+description: Common code for native_toolchain_rust and native_doctor.
+version: 0.1.0-dev.2
+homepage: https://github.com/irondash/native_toolchain_rust.git
+
+environment:
+  sdk: ^3.5.0-90.0.dev
```
Comment (mosuem) [inline native_toolchain_rust_common/pubspec.yaml:7]: `dart create` sets the minimum SDK version to the currently used one, I believe. Might make sense to relax this constraint here and elsewhere.

File: example/dart_package/example/dart_package_example.dart [hunk 13]
```diff
@@ -0,0 +1,6 @@
+import 'package:dart_package/dart_package.dart';
```
Comment (mosuem) [inline example/dart_package/example/dart_package_example.dart:1]: The folder would normally be called `bin/` for an application.

File: example/dart_package/pubspec.yaml [hunk 14]
```diff
@@ -0,0 +1,15 @@
+name: dart_package
+description: A starting point for Dart libraries or applications.
+version: 1.0.0
+publish_to: none
+
+environment:
+  sdk: ^3.5.0-90.0.dev
+
+dependencies:
+  native_toolchain_rust: ^0.1.0-dev.2
```
Comment (mosuem) [inline example/dart_package/pubspec.yaml:10]: As `native_toolchain_rust` is only used in `build.dart`, might as well be a dev dependency. Same for `native_assets_cli`. Also, for examples, I personally prefer them to be path dependencies on the package in the repo.

File: native_toolchain_rust_common/lib/src/command.dart [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+    stdoutEncoding: stdoutEncoding,
+  );
+  if (res.exitCode != 0) {
+    throw CommandFailedException(
+      executable: executable,
+      arguments: arguments,
+      result: res,
+    );
+  } else {
+    return res;
```
Comment (mosuem) [inline native_toolchain_rust_common/lib/src/command.dart:88]: Might as well return the `result.stdout as String`, given the usages and the fact that it throws on failure.

File: native_doctor/lib/src/toolchain_checker.dart [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+    required this.sections,
+    required this.proposedActions,
+  });
+}
+
+abstract class ActionLogger {
+  Future<void> logAction(String message, Future<void> Function() action);
+}
+
+abstract class ToolchainChecker {
```
Comment (mosuem) [inline native_doctor/lib/src/toolchain_checker.dart:54]: Some documentation on the fields here would be nice.

File: native_doctor/lib/src/native_doctor.dart [hunk 17]
```diff
@@ -0,0 +1,320 @@
...
+    writer.emptyLine();
+
+    final manifests = NativeManifest.forPackageConfig(
+      packageConfig,
+      verboseLogger: verboseLogger,
+    );
+
+    if (manifests.isEmpty) {
+      writer.printMessage(
+        'No dependency containing ${NativeManifest.fileName} found. Nothing to check.',
```
Comment (mosuem) [inline native_doctor/lib/src/native_doctor.dart:131]: This can be confusing, as no *other* packages were checked. But the `run native_doctor` did check the current package.

File: native_doctor/lib/src/checkers/ndk.dart [hunk 18]
```diff
@@ -0,0 +1,237 @@
...
+    _sdkInfo = await AndroidSdkInfo.find(
+      flutterRoot: doctor.flutterRoot?.toFilePath(),
+      logger: doctor.verboseLogger,
+    );
+
+    if (_sdkInfo == null) {
+      return ValidationResult(sections: [
+        ValidationResultSection(
+          ok: false,
+          title: 'Android SDK not found',
```
Comment (mosuem) [inline native_doctor/lib/src/checkers/ndk.dart:107]: Maybe add some link to https://developer.android.com/studio/index.html or similar?

File: native_toolchain_rust_common/lib/src/manifest.dart [hunk 19]
```diff
@@ -0,0 +1,118 @@
+import 'dart:io';
+
+import 'package:logging/logging.dart';
+import 'package:package_config/package_config.dart';
+import 'package:pub_semver/pub_semver.dart';
+import 'package:source_span/source_span.dart';
+import 'package:yaml/yaml.dart';
+
+class NdkManifestInfo {
+  NdkManifestInfo({
```
Comment (mosuem) [inline native_toolchain_rust_common/lib/src/manifest.dart:10]: Maybe use an extension type here?
```dart
extension type NdkManifestInfo(Version _version) {
```

File: native_doctor/lib/src/checkers/ndk.dart [hunk 20]
```diff
@@ -0,0 +1,237 @@
...
+
+    final result = <_Ndk>[];
+
+    final ndkDirs = await ndkDir.list().toList();
+    for (final dir in ndkDirs) {
+      final sourceProperties =
+          File.fromUri(dir.uri.resolve('source.properties'));
+      if (await sourceProperties.exists()) {
+        final content = await sourceProperties.readAsString();
+        final revision = content
```
Comment (mosuem) [inline native_doctor/lib/src/checkers/ndk.dart:72]: You could use a library such as https://pub.dev/packages/properties for that - but it's probably not worth it for one read.

File: native_doctor/lib/src/toolchain_checker.dart [hunk 21]
```diff
@@ -0,0 +1,65 @@
...
+  String get manifestKey;
+  String get displayName;
+
+  Future<void> updateFromManifest(
+    ManifestContext context,
+    YamlNode manifestNode,
+  );
+
+  Future<ValidationResult?> validate();
```
Comment (mosuem) [inline native_doctor/lib/src/toolchain_checker.dart:63]: Some documentation, such as was a `null` means here (as I understand that no check is needed?), would be helpful.

File: native_doctor/lib/src/writer.dart [hunk 22]
```diff
@@ -0,0 +1,190 @@
...
+    String message, {
+    String prefix = '',
+    TextColor? prefixColor,
+  });
+  void emptyLine();
+  bool get hasTerminal;
+  bool get supportsEmoji;
+}
+
+class AnsiWriter implements Writer, ActionLogger {
```
Comment (mosuem) [inline native_doctor/lib/src/writer.dart:33]: Again you might want to use a package such as https://pub.dev/packages/ansicolor for this - if you don't mind the additional dependency.

File: native_toolchain_rust/lib/src/android_linker_wrapper.dart [hunk 23]
```diff
@@ -0,0 +1,90 @@
...
+    let status = child.wait().unwrap_or_else(|err| {
+        eprintln!("cargokit (as linker): Failed to wait for {clang:?} to complete: {err}");
+        std::process::exit(1);
+    });
+
+    std::process::exit(status.code().unwrap_or(1))
+}
+''';
+
+class AndroidLinkerWrapper {
```
Comment (mosuem) [inline native_toolchain_rust/lib/src/android_linker_wrapper.dart:53]: What is the purpose of this class?

File: native_toolchain_rust/lib/src/android_environment.dart [hunk 24]
```diff
@@ -0,0 +1,167 @@
...
+  /// Toolchain being used.
+  final RustupToolchain toolchain;
+
+  /// Target being built.
+  final RustTarget target;
+
+  Future<Map<String, String>> buildEnvironment() async {
+    final toolchainPath = ndkInfo.toolchainPath;
+
+    final exe = Platform.isWindows ? '.exe' : '';
```
Comment (mosuem) [inline native_toolchain_rust/lib/src/android_environment.dart:89]: Maybe use `executableFileName` from `package:native_assets_cli`?

File: native_toolchain_rust/lib/src/builder.dart [hunk 25]
```diff
@@ -0,0 +1,275 @@
...
+  }
+
+  Future<void> _checkNativeManifest({
+    required BuildConfig buildConfig,
+  }) async {
+    final manifest = NativeManifest.forPackage(buildConfig.packageRoot);
+    if (manifest == null) {
+      throw RustToolchainException(
+        message: '`native_manifest.yaml` expected in package root.\n'
+            'See https://pub.dev/packages/native_doctor more information and example manifest.',
```
Comment (mosuem) [inline native_toolchain_rust/lib/src/builder.dart:84]:
```suggestion
            'See https://pub.dev/packages/native_doctor for more information and an example manifest.',
```

File: native_toolchain_rust/lib/src/builder.dart [hunk 26]
```diff
@@ -0,0 +1,275 @@
...
+
+  /// Build config provided to the build callback from `native_assets_cli`.
+  final BuildConfig buildConfig;
+
+  /// Dart build files inside hook directory that should be added as
+  /// dependencies. Default value adds `hook/build.dart` as dependency.
+  final List<String> dartBuildFiles;
+
+  /// By default `native_toolchain_rust` expects `native_manifest.yaml` in
+  /// package root in order to check for required Rust version and also for
```
Comment (mosuem) [inline native_toolchain_rust/lib/src/builder.dart:138]:
```suggestion
  /// package root in order to check for the required Rust version and also for
```

File: native_toolchain_rust/lib/src/builder.dart [hunk 27]
```diff
@@ -0,0 +1,275 @@
...
+  /// `native_doctor` to work. If you don't want to include `native_manifest.yaml`
+  /// in your package, set this to `true`.
+  ///
+  /// See https://pub.dev/packages/native_doctor for more information.
+  final bool ignoreMissingNativeManifest;
```
Comment (mosuem) [inline native_toolchain_rust/lib/src/builder.dart:143]: Maybe `useNativeManifest` do avoid double negatives in the check further below?

File: native_toolchain_rust/lib/src/builder.dart [hunk 28]
```diff
@@ -0,0 +1,275 @@
...
+      iosSdk: buildConfig.targetOS == OS.iOS ? buildConfig.targetIOSSdk : null,
+    )!;
+
+    await toolchain._checkTarget(target: target);
+    if (!ignoreMissingNativeManifest) {
+      await toolchain._checkNativeManifest(buildConfig: buildConfig);
+    }
+
+    if (!buildConfig.dryRun) {
+      await toolchain.toolchain.rustup.runCommand(
```
Comment (mosuem) [inline native_toolchain_rust/lib/src/builder.dart:184]: For my use case specifically, I probably want to pass some more arguments to `cargo`, see https://github.com/dart-lang/i18n/blob/main/.github/workflows/intl4x_artifacts.yml. Maybe adding a field `List<String> additionalFlags` to the `RustBuilder` might help? I don't know if the order of args matters - that would make things more complicated.

**Review by mosuem — COMMENTED** [review]
Really cool stuff!

You are running the tests for different platforms, might want to also test the lower SDK bounds, and check if `pub downgrade` breaks anything.

File: example/dart_package/hook/build.dart (hunk 2 above)
Comment (knopp) [reply example/dart_package/hook/build.dart:17]: It's not necessarily needed I just don't feel like printing stacktrace is particularly helpful, especially if the underlying exception is most likely caused by `cargo` invocation failing. In that case the stacktrace to running cargo command is just noise.

File: native_doctor/lib/src/native_doctor.dart (hunk 5 above)
Comment (knopp) [reply native_doctor/lib/src/native_doctor.dart:138]: That indeed would be the plan longer term. At this point this is more or less an experiment on how seamless we can make consuming a Rust dependency.

I called it native_doctor because apart from rustup it also checks and install NDK.

File: native_doctor/pubspec.lock (hunk 8 above)
Comment (knopp) [reply native_doctor/pubspec.lock:1]: Native doctor is meant to be used as tool, not a package, so I'd assume checking in `pubspec.lock` is the right thing to do?

File: native_toolchain_rust/lib/src/builder.dart (hunk 10 above)
Comment (knopp) [reply native_toolchain_rust/lib/src/builder.dart:24]: They only communicate through manifest and the manifest itself contain a version number.

File: native_toolchain_rust_common/pubspec.yaml (hunk 12 above)
Comment (knopp) [reply native_toolchain_rust_common/pubspec.yaml:7]: I didn't bother with this since native assets are still experimental and anyone using this will unlikely have dart version small than this.

File: example/dart_package/example/dart_package_example.dart (hunk 13 above)
Comment (knopp) [reply example/dart_package/example/dart_package_example.dart:1]: Makes sense. Not sure how it ended up with example, maybe I was looking at the flutter project.

File: example/dart_package/pubspec.yaml (hunk 14 above)
Comment (knopp) [reply example/dart_package/pubspec.yaml:10]: I'd prefer that, but even the example flutter ffi package has `native_toolchain_cli` and `native_toolchain_c` declared as dependencies. I'm wondering if maybe having them inside `dev_dependencies` causes problem when the package is a transient dependency? Or is there some other reason for that?

File: native_toolchain_rust_common/lib/src/manifest.dart (hunk 19 above)
Comment (knopp) [reply native_toolchain_rust_common/lib/src/manifest.dart:10]: I am a bit torn on this. Yes, both NdkManifestInfo and RustManifestInfo only have one field, but that's not intentional, it just happened to be that way. Maybe in future new fields are added, which means it will get reverted to a normal class, which to me indicates that maybe, conceptually, it's not really an extension type.

File: native_doctor/lib/src/writer.dart (hunk 22 above)
Comment (knopp) [reply native_doctor/lib/src/writer.dart:33]: It didn't want to have dependency on package with unverified publisher, but I see now that it actually lives in google repository? Any idea why it is unverified?

File: native_toolchain_rust/lib/src/android_linker_wrapper.dart (hunk 23 above)
Comment (knopp) [reply native_toolchain_rust/lib/src/android_linker_wrapper.dart:53]: I'll add the following comment:
```dart
/// This class is responsible for creating and building the linker wrapper.
/// This avoid using NDK wrapper scripts, which on windows cause problems when invoked
/// from rustc due to wrong CMD argument escaping.
///
/// https://github.com/android/ndk/issues/1856
```

File: native_toolchain_rust/lib/src/builder.dart (hunk 27 above)
Comment (knopp) [reply native_toolchain_rust/lib/src/builder.dart:143]: That seems better.

File: native_doctor/lib/src/toolchain_checker.dart (hunk 21 above)
Comment (knopp) [reply native_doctor/lib/src/toolchain_checker.dart:63]: Indeed `null` means no check was needed. I.e. for Rust it means that no package declared Rust requirement in the manifest. I'm not too happy with `null` being used here, it's been two weeks and I already had to look into the code to be sure :) 

I think I'll refactor `ValidationResult` to account for this instead.

File: native_toolchain_rust/lib/src/builder.dart (hunk 28 above)
Comment (knopp) [reply native_toolchain_rust/lib/src/builder.dart:184]: Extra cargo arguments are missing. This slipped my mind. I couldn't even port [super_native_extensions](https://github.com/superlistapp/super_native_extensions) (the primary use-case for us right now) to native assets without this.

File: example/dart_package/hook/build.dart (hunk 9 above)
Comment (knopp) [reply example/dart_package/hook/build.dart:11]: No particular reason. I think I'm fine with renaming it to `cratePath` and removing the `Cargo.toml` part from API.

File: example/dart_package/pubspec.yaml (hunk 14 above)
Comment (mosuem) [reply example/dart_package/pubspec.yaml:10]: @dcharkes

Comment (dcharkes) [reply example/dart_package/pubspec.yaml:10]: build hook and link hook dependencies cannot be dev dependencies. Details:

* https://github.com/dart-lang/native/issues/160

TL;DR: You don't want version skew between what version of helper packages where used to run the hooks and that are used at runtime later.

(Some more related discussions around managing dependencies and running hooks during the build: https://github.com/dart-lang/pub/issues/3794)

File: native_doctor/lib/src/writer.dart (hunk 22 above)
Comment (mosuem) [reply native_doctor/lib/src/writer.dart:33]: I am following up on this internally. Thanks for pointing that out!
